src/app/patron/add-edit-book-borrowed/add-edit-book-borrowed.spec.ts

```diff
@@ -0,0 +1,143 @@
+            }
+        ]));
+
+
+        TestBed.configureTestingModule({
+            declarations: [
+                AddEditBookBorrowedComponent
+            ],
+            imports: [
+                BrowserModule,
+                FormsModule,
+                ReactiveFormsModule,
+                RouterTestingModule,
+                HttpClientTestingModule
+            ],
+            providers: [
+                { provide: Router, useClass: class { navigate = jasmine.createSpy("navigate"); } },
+                { provide: AuthorService, useValue: [] },
+                {
+                    provide: ActivatedRoute, useValue: {
+                        snapshot: {
+                            params: {
+                                idBookBorrowed: '',
+                                idPatron: ''
+                            },
+                        },
+                    },
+                },
+                {provide: BookBorrowedService, useValue: mockBookBorrowedService},
+                { provide: BookService, useValue: mockBookService },
+                { provide: PatronService, useValue: mockPatronService }
+            ]
+        }).compileComponents().then(() => {
+            fixture = TestBed.createComponent(AddEditBookBorrowedComponent);
+            comp = fixture.componentInstance;
+            de = fixture.debugElement.query(By.css('form'));
+            el = de.nativeElement;
+            fixture.detectChanges();
+        });
+    }));
+
+    it('should create', () => {
+        expect(comp).toBeTruthy();
+    });
+
+    it(`should call the getBooks method`, () => {
+        expect(comp.books.length).toBeGreaterThan(0);
+    });
+
+    it(`should call the onSubmit method and set submitted to true`, () => {
+        comp.onSubmit();
+        expect(comp.submitted).toBeTruthy();
+    });
+
+    it(`should call the onAddBookBorrowed method`, () => {
+        spyOn(comp, 'onAddBookBorrowed');
+        comp.form.controls['book'].setValue({
+            id: 1, title: 'title 1', publishedDate: '11/10/2021', quantity: 1, description: 'description 1', author: { "id": 2, "firstName": "Oscar", "lastName": "Wilde" },
+            isbns: [{ id: 1, borrowed: false, book_id: 1, isbn: "isbn1 carte1" },
+            { id: 2, borrowed: false, book_id: 1, isbn: "isbn2 carte1" }], genres: [{ id: 1, name: "Comedy" }]
+        });
+        comp.form.controls['isbn'].setValue({ id: 2, borrowed: false, book_id: 1, isbn: "isbn2 carte1" });
+        comp.onSubmit();
+        expect(comp.onAddBookBorrowed).toHaveBeenCalled();
+    });
+
+
+
+    it(`form should be invalid`, () => {
+        comp.form.controls['book'].setValue('');
+        comp.form.controls['isbn'].setValue('');
+
+        expect(comp.form.valid).toBeFalsy();
+
+    });
+
+    it(`form should be valid`, () => {
+        comp.form.controls['book'].setValue({
```

Comment:
We can edit or create another test and mock the `  tbookService.getBookById` instead of setting the form values manually in case of edit book.